Pass deeply nested keys correctly to afterSave triggers

mstniy opened the pull request:
### New Pull Request Checklist
<!--
    Please check the following boxes [x] before submitting your issue.
    Click the "Preview" tab for better readability.
    Thanks for contributing to Parse Server!
-->

- [x] I am not disclosing a [vulnerability](https://github.com/parse-community/parse-server/blob/master/SECURITY.md).
- [x] I am creating this PR in reference to an [issue](https://github.com/parse-community/parse-server/issues?q=is%3Aissue).

### Issue Description
<!-- Add a brief description of the issue this PR solves. -->

Related issue: #7384 

### Approach
RestWrite.buildUpdatedObject now functions correctly even for deeply nested keys.

### TODOs before merging
<!--
    Add TODOs that need to be completed before merging this PR.
    Delete suggested TODOs that do not apply to this PR.
-->

- [x] Add test cases
- [x] Add entry to changelog
- [ ] Add changes to documentation (guides, repository pages, in-code descriptions)
- [ ] Add [security check](https://github.com/parse-community/parse-server/blob/master/CONTRIBUTING.md#security-checks)
- [ ] Add new Parse Error codes to Parse JS SDK <!-- no hard-coded error codes in Parse Server -->
- [ ] ...

mstniy committed:
Pass deeply nested keys correctly to afterSave triggers
  Also added a regression test

mstniy committed:
Simplified the approach

dplewis reviewed (dismissed):
Nice cleanup and test!

dplewis commented:
@mstniy Can you add this fix to the ChangeLog?

mtrezza commented:
This looks suspicious.

This PR removes lot of code from RestWrite and replaces it with a line that was already within the removed code.

Did you research why the logic around the line that you intend to keep was there previously? Its possible that the logic was there to accommodate certain use cases.

mstniy commented:
@mtrezza ParseObject.set is capable of handling nested keys, no matter the depth. The regression tests actually makes use of this capability. I am not sure why the previous snippet did the split manually. Perhaps it was written before ParseObject.set gained that ability, or the contributor wasn't aware it could do that? But that's pure speculation.

Thus, the else branch also becomes a simple call to ParseObject.set, so I have removed the branch entirely. I am not sure why the code checks for a dot, so I have left that part intact.

mstniy committed:
Added changelog entry

mstniy commented:
> @mstniy Can you add this fix to the ChangeLog?

Done

dplewis commented:
@mstniy @mtrezza Support for nested objects was recently improved in the SDK 
https://github.com/parse-community/Parse-SDK-JS/releases/tag/3.1.0

That’s the reason why this cleanup works so well.

mtrezza commented:
Aha, that may explain it indeed, thanks.

mstniy committed:
Expanded the regression test

mstniy committed:
Use a patched version of the js sdk
  Solves problems with nested keys
Expanded the regression test

mstniy committed:
Modified a test
  ParseUser.get('password') used to be undefined.
  Now, password is not an attribute

mstniy committed:
Merge branch 'master' into deep_7384

mstniy committed:
Updated the sdk fork

mstniy commented:
Blocked on parse-community/Parse-SDK-JS#1364

mstniy committed:
Remove console.log

mtrezza commented:
> ## ⚠️ Important change for merging PRs from Parse Server 5.0 onwards!
  >
  > We are planning to release the first beta version of Parse Server 5.0 in October 2021.
  >
  > If a PR contains a breaking change and is not merged before the _beta_ release of Parse Server 5.0, it cannot be merged until the end of 2022. Instead it has to follow the [Deprecation Policy](https://github.com/parse-community/parse-server/blob/master/CONTRIBUTING.md#deprecation-policy) and phase-in breaking changes to be merged during the course of 2022.
  >
  > One of the most voiced community feedbacks was the demand for predictability in breaking changes to make it easy to upgrade Parse Server. We have made a first step towards this by introducing the Deprecation Policy in February 2021 that assists to phase-in breaking changes, giving developers time to adapt. We will follow-up with the introduction of [Release Automation](https://github.com/parse-community/parse-server/issues/7271) and a branch model that will allow breaking changes only with a new major release, scheduled for the beginning of each calendar year.
  >
  > We understand that some PRs are a long time in the making and we very much appreciate your contribution. We want to make it easy for PRs that contain a breaking change and were created _before_ the introduction of the Deprecation Policy. These PRs can be merged with a breaking change without being phased-in before the beta release of Parse Server 5.0. We are making this exception because we appreciate that this is a time of transition that requires additional effort from contributors to adapt. We encourage everyone to prepare their PRs until the end of September and account for review time and possible adaptions.
  > 
  > If a PR contains a breaking change and should be merged before the beta release, please mention @parse-community/server-maintenance and we will coordinate with you to merge the PR.
  >
  > *Thanks for your contribution and support during this transition to Parse Server release automation!*

RahulLanjewar93 commented:
Hey, just found that this issue still persist with `parse-server@7.5.2` and `parse@5.3.0`. 
If the pr looks good, can someone merge and release it? 
Thanks !

dplewis commented:
@RahulLanjewar93 I thought @mstniy fixed this on the JavaScript client side. Can you open a new issue?

Edit: Oh I see what they did, I basically copied the dot notation JS code to the server before they improved it

RahulLanjewar93 commented:
> @RahulLanjewar93 I thought @mstniy fixed this on the JavaScript client side. Can you open a new issue?
> 
> Edit: Oh I see what they did, I basically copied the dot notation JS code to the server before they improved it

I think we still need this pr to be merged so that it uses the default set behaviour from parse js sdk?

dplewis commented:
@RahulLanjewar93 Can you copy the code and open a new PR? This one is on draft mode

RahulLanjewar93 commented:
> @RahulLanjewar93 Can you copy the code and open a new PR? This one is on draft mode

Sure

RahulLanjewar93 commented:
@dplewis created a new pr here by copying the changes. 
Should it be pointed to alpha or release-7.x.x? 
#9720
